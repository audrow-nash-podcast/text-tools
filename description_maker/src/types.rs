#[derive(Clone, Debug)]
pub struct Link {
    pub text: String,
    pub href: String,
}

#[derive(Clone, Debug)]
pub struct Timestamp {
    pub hours: u16,
    pub minutes: u16,
    pub seconds: u16,
}

impl std::fmt::Display for Timestamp {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{:02}:{:02}:{:02}",
            self.hours, self.minutes, self.seconds
        )
    }
}

#[derive(Clone, Debug)]
pub struct TimeCode {
    pub text: String,
    pub timestamp: Timestamp,
}

#[derive(Clone, Debug)]
pub struct Episode {
    pub title: String,
    pub number: u16,
    pub description: String,
    pub links: Vec<Link>,
    pub time_codes: Vec<TimeCode>,
}

#[derive(Clone, Debug)]
pub struct PodcastInfo {
    pub name: String,
    pub transcript_site_url: String,
    pub links: Vec<Link>,
}
