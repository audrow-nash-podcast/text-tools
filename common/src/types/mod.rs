pub mod time_code;
pub use crate::types::time_code::TimeCode;

pub mod outline_entry;
pub use crate::types::outline_entry::OutlineEntry;
