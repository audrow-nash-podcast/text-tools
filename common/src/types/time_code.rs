use regex::Regex;
use std::cmp::Ordering;
use std::error::Error;
use std::str::FromStr;

#[derive(Debug, PartialEq)]
pub struct TimeCode {
    pub hours: u32,
    pub minutes: u32,
    pub seconds: u32,
}

impl PartialOrd for TimeCode {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        if self.hours > other.hours {
            Some(Ordering::Greater)
        } else if self.hours < other.hours {
            Some(Ordering::Less)
        } else if self.minutes > other.minutes {
            Some(Ordering::Greater)
        } else if self.minutes < other.minutes {
            Some(Ordering::Less)
        } else if self.seconds > other.seconds {
            Some(Ordering::Greater)
        } else if self.seconds < other.seconds {
            Some(Ordering::Less)
        } else {
            Some(Ordering::Equal)
        }
    }
}

impl TimeCode {
    pub fn from_str(text: &str) -> Result<TimeCode, Box<dyn Error>> {
        let time_code_regex =
            Regex::new(r"(\d{2,}):(\d{2}):(\d{2})").expect("Time code regex is valid");
        let captures = time_code_regex.captures(text).ok_or("Invalid time code")?;

        let hours = u32::from_str(&captures[1])?;
        let minutes = u32::from_str(&captures[2])?;
        let seconds = u32::from_str(&captures[3])?;

        if minutes > 59 {
            return Err("Minutes must be between 0-59".into());
        }
        if seconds > 59 {
            return Err("Seconds must be between 0-59".into());
        }

        return Ok(TimeCode {
            hours,
            minutes,
            seconds,
        });
    }
}

impl std::fmt::Display for TimeCode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{:02}:{:02}:{:02}",
            self.hours, self.minutes, self.seconds
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn from_str() {
        assert_eq!(
            TimeCode::from_str("00:00:00").expect("Time code should parse"),
            TimeCode {
                hours: 0,
                minutes: 0,
                seconds: 0,
            }
        );
        assert_eq!(
            TimeCode::from_str("00:00:01").expect("Time code should parse"),
            TimeCode {
                hours: 0,
                minutes: 0,
                seconds: 1,
            }
        );
        assert_eq!(
            TimeCode::from_str("00:01:00").expect("Time code should parse"),
            TimeCode {
                hours: 0,
                minutes: 1,
                seconds: 0,
            }
        );
        assert_eq!(
            TimeCode::from_str("01:00:00").expect("Time code should parse"),
            TimeCode {
                hours: 1,
                minutes: 0,
                seconds: 0,
            }
        );
        assert_eq!(
            TimeCode::from_str("01:01:01").expect("Time code should parse"),
            TimeCode {
                hours: 1,
                minutes: 1,
                seconds: 1,
            }
        );
    }

    #[test]
    fn equality() {
        let t1 = TimeCode {
            hours: 1,
            minutes: 2,
            seconds: 3,
        };

        let t2 = TimeCode {
            hours: 1,
            minutes: 2,
            seconds: 3,
        };

        assert_eq!(t1, t2);

        let t3 = TimeCode {
            hours: 1,
            minutes: 2,
            seconds: 4,
        };

        assert!(t1 != t3)
    }

    #[test]
    fn order_time_codes() {
        let t1 = TimeCode {
            hours: 0,
            minutes: 0,
            seconds: 0,
        };

        let t2 = TimeCode {
            hours: 0,
            minutes: 0,
            seconds: 1,
        };

        let t3 = TimeCode {
            hours: 0,
            minutes: 1,
            seconds: 0,
        };

        let t4 = TimeCode {
            hours: 1,
            minutes: 0,
            seconds: 0,
        };

        let t5 = TimeCode {
            hours: 1,
            minutes: 1,
            seconds: 1,
        };

        assert!(t1 < t2);
        assert!(t2 < t3);
        assert!(t3 < t4);
        assert!(t4 < t5);
    }
}
