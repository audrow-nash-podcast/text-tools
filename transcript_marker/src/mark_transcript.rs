use crate::types::{OutlineEntry, TimeCode};
use regex::Regex;
use std::error::Error;

pub fn mark_transcript(
    text: &str,
    outline_entries: &mut Vec<OutlineEntry>,
) -> Result<String, Box<dyn Error>> {
    let line_regex = Regex::new(r"\[(\d{2,}:\d{2}:\d{2})\] .*").expect("Line regex is valid");

    outline_entries.sort();

    let mut output_text: Vec<String> = vec![];
    output_text.push("## Table of Contents\n".into());
    output_text.push(get_md_table_of_contents(&outline_entries) + "\n");
    for line in text.lines() {
        if let Some(captures) = line_regex.captures(line) {
            let time_code = TimeCode::from_str(&captures[1])?;
            if outline_entries.len() > 0 && time_code >= outline_entries[0].time_code {
                let entry = outline_entries.remove(0);
                output_text.push(format!("## {}\n", entry.text));
            }
        }
        output_text.push(format!("{}", line.trim()));
    }

    Ok(output_text.join("\n"))
}

fn get_md_table_of_contents(outline_entries: &Vec<OutlineEntry>) -> String {
    let mut output_text: Vec<String> = vec![];
    for entry in outline_entries {
        output_text.push(format!(
            "- [[{}] {}]({})",
            entry.time_code,
            entry.text,
            get_md_heading_url(&entry.text)
        ));
    }
    output_text.join("\n")
}

fn get_md_heading_url(text: &str) -> String {
    format!(
        "#{}",
        text.to_lowercase()
            .replace(" ", "-")
            .chars()
            .filter(|c| c.is_alphanumeric() || *c == '-')
            .collect::<String>()
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;

    #[test]
    fn put_text_before_timecodes() {
        let mut outline_entries: Vec<OutlineEntry> = vec![
            OutlineEntry {
                time_code: TimeCode {
                    hours: 0,
                    minutes: 0,
                    seconds: 0,
                },
                text: "Introduction".into(),
            },
            OutlineEntry {
                time_code: TimeCode {
                    hours: 0,
                    minutes: 1,
                    seconds: 30,
                },
                text: "Nag and Mike introduce themselves".into(),
            },
            OutlineEntry {
                time_code: TimeCode {
                    hours: 0,
                    minutes: 3,
                    seconds: 09,
                },
                text: "Nag and Mike introduce electric sheep".into(),
            },
        ];

        let transcript = r#"[00:00:00] **Audrow Nash:** I've talked to a lot of people about this interview, and I'm excited that I get to share it with you. In it, I talk with Nag and Mike from Electric Sheep. They're doing many things differently than most robotics companies that I've talked to, and they're making big bets that I think will pay off. Here are three examples to show you what I mean.

First, they're throwing away classical robotics approaches, and instead... using machine learning. I'm not just talking about for perception or parameter optimization, but even for things like localization or high level control. Second, they've turned the lawn mowing problem on its head to make robots that are intrinsically safer.

And third, instead of selling their robots or doing a subscription model, they buy profitable landscaping businesses and give those companies robots. There are a lot of advantages to this last point and you'll see the details of each during the interview. I hope it surprises you as much as it did me. I think you'll enjoy this interview if you're curious about how AI and robotics can fit together in a real application and if you want to see a new robotics business model that I think will be very popular in the near future.

[00:01:37] **Nag Murty:** Yeah. Hi everyone. I'm Nag Murthy. I'm the CEO and co founder of Electric Sheep Robotics. 

And then over time, we plan to use those, automated robots to improve our own margins over time. So that's what we do. 

[00:03:09] **Michael Laskey:** Yeah, I can take that one."#;

        let new_transcript = mark_transcript(transcript, &mut outline_entries)
            .expect("Mark transcript should succeed");
        assert_snapshot!(new_transcript);
    }

    #[test]
    fn get_md_heading_url_test() {
        assert_eq!(get_md_heading_url("Start"), "#start".to_string());
        assert_eq!(
            get_md_heading_url("Introducing Bradley + Luxonis"),
            "#introducing-bradley--luxonis".to_string()
        );
        assert_eq!(
            get_md_heading_url("Special characters!@#$%^&*()_+[]~><👍🤖"),
            "#special-characters".to_string()
        );
    }
}
