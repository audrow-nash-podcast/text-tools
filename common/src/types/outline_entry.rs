use crate::types::time_code::TimeCode;
use serde::{Deserialize, Serialize};
use std::cmp::Ordering;

#[derive(Debug, PartialEq, Serialize, Deserialize, Clone)]
pub struct OutlineEntry {
    pub time_code: TimeCode,
    pub text: String,
}

impl Eq for OutlineEntry {}

impl Ord for OutlineEntry {
    fn cmp(&self, other: &Self) -> Ordering {
        if self.time_code > other.time_code {
            Ordering::Greater
        } else if self.time_code < other.time_code {
            Ordering::Less
        } else {
            Ordering::Equal
        }
    }
}

impl PartialOrd for OutlineEntry {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}
