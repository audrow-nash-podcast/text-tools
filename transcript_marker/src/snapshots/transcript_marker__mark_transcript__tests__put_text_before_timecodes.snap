---
source: transcript_marker/src/mark_transcript.rs
expression: new_transcript
---
## Table of Contents

- [[00:00:00] Introduction](#introduction)
- [[00:01:30] Nag and Mike introduce themselves](#nag-and-mike-introduce-themselves)
- [[00:03:09] Nag and Mike introduce electric sheep](#nag-and-mike-introduce-electric-sheep)

## Introduction

[00:00:00] **Audrow Nash:** I've talked to a lot of people about this interview, and I'm excited that I get to share it with you. In it, I talk with Nag and Mike from Electric Sheep. They're doing many things differently than most robotics companies that I've talked to, and they're making big bets that I think will pay off. Here are three examples to show you what I mean.

First, they're throwing away classical robotics approaches, and instead... using machine learning. I'm not just talking about for perception or parameter optimization, but even for things like localization or high level control. Second, they've turned the lawn mowing problem on its head to make robots that are intrinsically safer.

And third, instead of selling their robots or doing a subscription model, they buy profitable landscaping businesses and give those companies robots. There are a lot of advantages to this last point and you'll see the details of each during the interview. I hope it surprises you as much as it did me. I think you'll enjoy this interview if you're curious about how AI and robotics can fit together in a real application and if you want to see a new robotics business model that I think will be very popular in the near future.

## Nag and Mike introduce themselves

[00:01:37] **Nag Murty:** Yeah. Hi everyone. I'm Nag Murthy. I'm the CEO and co founder of Electric Sheep Robotics.

And then over time, we plan to use those, automated robots to improve our own margins over time. So that's what we do.

## Nag and Mike introduce electric sheep

[00:03:09] **Michael Laskey:** Yeah, I can take that one.
